assets/blocks/course-outline/data.js

```diff
@@ -71,9 +71,11 @@ export const syncStructureToBlocks = ( structure, blocks ) => {
  * @param {[CourseLessonData,CourseModuleData]} item   Structure item.
  * @return {Object} The block, if found.
  */
-const findBlock = ( blocks, { id, type } ) => {
+const findBlock = ( blocks, { id, type, title } ) => {
 	const compare = ( { name, attributes } ) =>
-		id === attributes.id && blockNames[ type ] === name;
+		( id === attributes.id ||
+			( ! attributes.id && attributes.title === title ) ) &&
```

Comment:
Thinking about it, it should work out: this finds the first block with the same title, which has the ID attribute set before looking for the next one. So the second call for this will find the second block. The blocks should also be in the correct order here most of the time. (But worst case scenario is the editor switches them)